### PR TITLE
[SPARK-48379][INFRA] Cancel previous builds on a PR when new commit is pushed

### PR DESCRIPTION
**Pull request by stefankandic**
### What changes were proposed in this pull request?

When there is an open PR with a build in progress and a new commit is pushed I propose that we should cancel the previous build.

We would only want to do this on forked repositories.

[Concurrecy docs](https://docs.github.com/en/actions/using-jobs/using-concurrency)

### Why are the changes needed?

To reduce wait times for newer commit's and save compute resources.

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

Manually.

### Was this patch authored or co-authored using generative AI tooling?

No.

### CODE REVIEW

**Review by HyukjinKwon — APPROVED** [review]
Nice, LGTM if it works

**Comment by stefankandic** [comment]
@HyukjinKwon Cancellation works fine, but the issue is that `Notify test workflow` is sometimes failing with `Failed to retrieve check_run_head after 3 attempts`.

Not sure, what is causing this, maybe we should wait a bit longer for a workflow to be started or something? Tried increasing the sleep time in the retry but it still seems to use the old value of 3 seconds:
image

Do you have any ideas as to why this is happening as I see that you worked on this as well as @yaooqinn

File: .github/workflows/build_main.yml [hunk 1]
```diff
@@ -24,6 +24,10 @@ on:
     branches:
     - '**'
 
+concurrency:
+  group: ${{ github.workflow }}-${{ github.ref }}
+  cancel-in-progress: ${{ github.ref != 'refs/heads/master' && !startsWith(github.ref, 'refs/heads/branch-') }}
```
Comment (HyukjinKwon) [inline .github/workflows/build_main.yml:29]: I think we should do this only in forked repository btw.

**Comment by HyukjinKwon** [comment]
I retriggered, and seems working fine. I think it just failed because it tried to get the test run of the cancelled build.

File: .github/workflows/build_main.yml (hunk 1 above)
Comment (HyukjinKwon) [reply .github/workflows/build_main.yml:29]: So we shouldn't cancel them if we're in apache spark repo.

**Comment by HyukjinKwon** [comment]
We should also probably remove https://github.com/apache/spark/actions/workflows/cancel_duplicate_workflow_runs.yml and replace that to the approach used here (but of course should be done in a separate PR).

**Comment by yaooqinn** [comment]
I think the reason could be either due to cancellation or exceeding the per_page results, instead of a timeout.

**Comment by HyukjinKwon** [comment]
hmmmm

**Comment by HyukjinKwon** [comment]
I think this is fine.

**Comment by HyukjinKwon** [comment]
Merged to master.

**Comment by yaooqinn** [comment]
> @yaooqinn Do you know what would be the workaround?

The only suitable way seems to increase the await time for a new job to be scheduled

File: .github/workflows/build_main.yml [hunk 2]
```diff
@@ -24,6 +24,11 @@ on:
     branches:
     - '**'
 
+concurrency:
+  group: ${{ github.workflow }}-${{ github.ref }}
+  # only cancel in forked repositories
+  cancel-in-progress: ${{ github.repository_owner != 'apache' }}
```
Comment (HyukjinKwon) [inline .github/workflows/build_main.yml:30]: Okay, there are two issues :-).

1. With `concurrency`, it introduces a new state like pending:



and the notify job fails at https://github.com/apache/spark/blob/master/.github/workflows/notify_test_workflow.yml#L122

So increasing the timeout here a lot would help.

2. The builds in the master branch in this main repo became sequential. It doesn't cancel and wait for the build in the previous commit to finish.

**Comment by HyukjinKwon** [comment]
Let me revert it out for now.

**Pull request closed without merging**